### ISSUE TITLE
JSON Schema draft 4 not parsed?

### ISSUE THREAD

**Issue by dave-irvine** [raml-org/raml-js-parser-2#528]
If I add a v4 JSON Schema as a type to my API and then dump out my API as JSON, it seems like the schema does not get loaded. Am I doing something wrong, or is draft 4 not supported yet?

```
types: Array[1]
  0: Object
    Person: Object
        __METADATA__: Object
       displayName: "Person"
       name: "Person"
       required: true
       type: Array[1]
           0:"{↵  "$schema": "http://json-schema.org/draft-04/schema#",↵  "title": "Person",↵.......
```

**Comment by ddenisenko** [comment]
I dont think its about the version of JSON schema.

Could you please provide a complete RAML example. If the schema is an external file, please also attach it.

**Comment by dave-irvine** [comment]
## Test.js:

```
var raml1Parser = require('raml-1-parser');
var util = require('util');
var api = raml1Parser.loadApiSync("./person.raml");

console.log(util.inspect(api.toJSON(), {depth:null}));
```

## person.raml:
```
#%RAML 1.0
title: My API With Types
types:
  Person: !include person.json
```

## person.json:
````
{
  "title": "Example Schema",
  "type": "object",
  "properties": {
    "firstName": {
      "type": "string"
    },
    "lastName": {
      "type": "string"
    }
  },
  "required": ["firstName", "lastName"]
}
````

### output:

````
{ types: 
   [ { Person: 
        { name: 'Person',
          displayName: 'Person',
          schemaPath: 'person.json',
          type: [ '{\n\t"title": "Example Schema",\n\t"type": "object",\n\t"properties": {\n\t\t"firstName": {\n\t\t\t"type": "string"\n\t\t},\n\t\t"lastName": {\n\t\t\t"type": "string"\n\t\t}\n\t},\n\t"required": ["firstName", "lastName"]\n}\n' ],
          required: true,
          __METADATA__: 
           { primitiveValuesMeta: 
              { displayName: { calculated: true },
                required: { insertedAsDefault: true } } } } } ],
  title: 'My API With Types' }
````

**Comment by dave-irvine** [comment]
@ddenisenko Any thoughts?

**Comment by ddenisenko** [comment]
@dave-irvine so, what's wrong with the output that you see? External type (schema) looks to be loaded fine. Do you expect it to be automatically converted to an object type?

**Comment by dave-irvine** [comment]
Yeah I kind of did actually :( When I read the spec I saw:

> A RAML processor MUST be able to interpret and apply JSON Schema and XML Schema.

Which I assumed meant I would get a useable Object out of the processor. Obviously if you use RAML types directly you get:

## output:
```
{ types: 
   [ { Person: 
        { name: 'Person',
          displayName: 'Person',
          type: [ 'object' ],
          required: true,
          properties: 
           { firstName: 
              { name: 'firstName',
                displayName: 'firstName',
                type: [ 'string' ],
                required: true,
                __METADATA__: 
                 { primitiveValuesMeta: 
                    { displayName: { calculated: true },
                      type: { insertedAsDefault: true },
                      required: { insertedAsDefault: true } } } },
             lastName: 
              { name: 'lastName',
                displayName: 'lastName',
                type: [ 'string' ],
                required: true,
                __METADATA__: 
                 { primitiveValuesMeta: 
                    { displayName: { calculated: true },
                      type: { insertedAsDefault: true },
                      required: { insertedAsDefault: true } } } } },
          __METADATA__: 
           { primitiveValuesMeta: 
              { displayName: { calculated: true },
                type: { insertedAsDefault: true },
                required: { insertedAsDefault: true } } } } } ],
  title: 'My API With Types' }
```

which is useable.

If I've misunderstood the spec, it might be useful to get a property on the returned objects that identify them as "external" types or something similar, so that I can know I need to pass that object to something that will parse the JSON schema for me.

**Comment by ddenisenko** [comment]
There are discussions regard the addition of the type's type to the JSON so that users could easily identify what they have at hands. For now please check for `schemaPath` to find out that is the schema.

@sichvoge @KonstantinSviridov here is another case of the need for type's type, so lets finally add it to the new JSON format.

**Comment by dave-irvine** [comment]
I don't really want to push on this, but if the spec says:

> A RAML processor MUST be able to interpret and apply JSON Schema and XML Schema.

Then surely this processor needs to parse the JSON Schema? Else it is not "interpreting" it, and couldn't apply it.

Please let me know if I am misunderstanding the spec.

**Comment by ddenisenko** [comment]
You are not misunderstanding the spec, the parser does load external types (JSON schema and XSD), interpret (validate) and apply them to the available XML and JSON content (examples). 

What it does not do is conversion from the external type to the object type, which the spec nowhere obligates the parser to do. 

There is the whole section of https://github.com/raml-org/raml-spec/blob/master/versions/raml-10/raml-10.md#using-xml-and-json-schema where the 'A RAML processor MUST be able to interpret and apply JSON Schema and XML Schema' phrase is coming from, which over and over repeats how external types can not be used in a mix with other types, whether it is inheritance, type expressions etc etc. That is exactly due to the external types not being converted to the normal object types, and this comes from the complexity and variety of such conversion.

**Comment by ddenisenko** [comment]
As was intended above, now users should have easier time with `type` property: there is the new `typePropertyKind`, which appears everywhere where `type` is mentioned and having one of the following values:

- `TYPE_EXPRESSION` in case of a type expression value of `type` property, including simple type name.
- `INPLACE` if `type` property value is an object describing inplace type definition.
- `JSON` if this is JSON-based external type
- `XML` if this is XML-based external type

Will be released in 1.1.11
